### PR TITLE
feat(data): introduce CedarValueMapper and validation enhancements

### LINKED ISSUE #13065: feat(jans-cedarling): Cedar Value Mapping and Validation
## Is your feature request related to a problem? Please describe.
The DataStore and policy evaluation pipeline require a reliable way to convert external JSON data into strongly typed `CedarValue`s and back again.  
This issue introduces bidirectional JSON ↔ CedarValue mapping with full type support, schema-aware validation, and robust error reporting to ensure correctness and security.

### Cedar Value Mapping
- [ ] Implement JSON → `CedarValue` mapping function  
- [ ] Support all Cedar data types:
  - Primitive types: `String`, `Long`, `Bool`
  - Collection types: `Set`, `Record`
  - Entity references (with schema-based validation)
  - Extension types (e.g. IP addresses, decimals)
- [ ] Implement reverse mapping (`CedarValue` → JSON)  
- [ ] Handle nested data structures  
- [ ] Support dot-notation access for nested records  

### Validation and Type Safety
- [ ] Implement data validation against Cedar schema constraints  
- [ ] Add strict type safety checks for all pushed data  
- [ ] Validate data size limits and format constraints  
- [ ] Add input sanitization to prevent malformed or unsafe data  
- [ ] Implement detailed validation error reporting  

## Acceptance Criteria
- [ ] JSON values are correctly mapped to corresponding `CedarValue` representations  
- [ ] All supported Cedar data types (primitive, collection, entity, extension) are handled correctly  
- [ ] Reverse mapping from `CedarValue` to JSON preserves structure and semantics  
- [ ] Nested structures and dot-notation access work as expected  
- [ ] Data is validated against Cedar schema constraints before being accepted  
- [ ] Invalid types, formats, or sizes are rejected with clear and actionable error messages  
- [ ] Input sanitization prevents malformed or unsafe data from entering the system  
- [ ] Validation and mapping logic is covered by unit tests and edge cases

### PR DESCRIPTION
**Pull request by haileyesus2433**
### Prepare

- [x] Read [PR guidelines](https://github.com/JanssenProject/jans/blob/main/docs/CONTRIBUTING.md#prs)
- [x] Read [license information](https://github.com/JanssenProject/jans/blob/main/LICENSE)

-------------------

#### Target issue
 closes #13065
-------------------
### Test and Document the changes
- [x] Static code analysis has been run locally and issues have been fixed
- [x] Relevant unit and integration tests have been added/updated
- [ ] Relevant documentation has been updated if any (i.e. user guides, installation and configuration guides, technical design docs etc)

Please check the below before submitting your PR. The PR will not be merged if there are no commits that start with `docs:` to indicate documentation changes or if the below checklist is not selected.
- [x] **I confirm that there is no impact on the docs due to the code changes in this PR.**

## Summary by CodeRabbit

* **New Features**
  * Added configurable JSON data validation with constraints for depth, string/array/object sizes
  * Added bidirectional JSON-to-Cedar value conversion with automatic type detection and extension support
  * Enhanced error handling with detailed validation error reporting and error aggregation

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by mo-auto** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/janssenproject/pr-checks/4b2873dc-8b92-42f1-9b5e-791731b918bc?refs=6e484dcdd14fabd6a061ec7ab298f70a12dcf2e4&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

**Comment by haileyesus2433** [comment]
@coderabbitai review

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+    /// Create a new mapper with default settings.
+    pub fn new() -> Self {
+        Self {
+            auto_detect_extensions: true,
+            max_value_size: 0,
+        }
+    }
+
+    /// Create a mapper with auto-detection of extension types disabled.
+    pub fn without_auto_detect() -> Self {
```
Comment (olehbozhok) [inline jans-cedarling/cedarling/src/data/mapper.rs]: Maybe better to rename `new_without_auto_detect`?

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Detect if a string value represents a Cedar extension type.
+    pub fn detect_extension(value: &str) -> Option<ExtensionValue> {
+        // Check for IP address
+        if IpAddr::from_str(value).is_ok() {
```
Comment (olehbozhok) [inline jans-cedarling/cedarling/src/data/mapper.rs:238]: Cedar policy has more than one extension type:
https://docs.cedarpolicy.com/policies/syntax-datatypes.html#datatype-extension
Should we check them all?

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+            false
+        }
+    }
+
+    /// Parse an entity reference from JSON.
+    ///
+    /// # Returns
+    ///
+    /// A tuple of (entity_type, entity_id) if valid.
+    pub fn parse_entity_reference(value: &Value) -> Result<(String, String), ValueMappingError> {
```
Comment (olehbozhok) [inline jans-cedarling/cedarling/src/data/mapper.rs]: Instead of returning `(String, String)`, maybe better to create struture that has named attributes?

File: jans-cedarling/cedarling/src/data/mapper.rs (hunk 2 above)
Comment (olehbozhok) [reply jans-cedarling/cedarling/src/data/mapper.rs:238]: or add comment

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+                    return Err(ValueMappingError::CollectionErrors(errors));
+                }
+
+                RestrictedExpression::new_set(exprs)
+            },
+            Value::Object(obj) => {
+                // Check for entity reference
+                if Self::is_entity_reference(value) {
+                    let (entity_type, entity_id) = Self::parse_entity_reference(value)?;
+                    let uid_str = format!("{}::\"{}\"", entity_type, entity_id);
```
Comment (olehbozhok) [inline jans-cedarling/cedarling/src/data/mapper.rs]: Please don't use string concatenation for parsing uid.
Better to parse `EntityId` and `EntityTypeName` separately.

this [PR](https://github.com/JanssenProject/jans/pull/12988/changes) dramatically improved perfomance.

File: jans-cedarling/cedarling/src/data/validation.rs [hunk 5]
```diff
@@ -0,0 +1,769 @@
...
+                    .collect();
+                Value::Object(sanitized)
+            },
+            // Numbers and booleans pass through unchanged
+            _ => value.clone(),
+        }
+    }
+
+    // Internal validation at a specific path and depth
+    fn validate_at_path(
```
Comment (olehbozhok) [inline jans-cedarling/cedarling/src/data/validation.rs:339]: We are going to apply `pedantic` clippy linter. It would be great to have this function with fewer than 100 rows.

File: jans-cedarling/cedarling/src/data/validation.rs [hunk 6]
```diff
@@ -0,0 +1,769 @@
...
+            Err(ValidationError::TypeMismatch {
+                path: "$".to_string(),
+                expected: format!("{:?}", expected),
+                actual: format!("{:?}", actual),
+            })
+        }
+    }
+
+    /// Validate an extension value format.
+    pub fn validate_extension(
```
Comment (olehbozhok) [inline jans-cedarling/cedarling/src/data/validation.rs:232]: This method doesn't use `self`, so we don't need to have it as an argument. Same to others.

**Comment by olehbozhok** [comment]
@coderabbitai if the method has `&self` as an argument but doesn't use it. This argument should be deleted.

File: jans-cedarling/cedarling/src/data/mapper.rs (hunk 4 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/92d67b958db185aae5bf0466744695d37eec4285

File: jans-cedarling/cedarling/src/data/validation.rs (hunk 5 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/validation.rs:339]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/ebd30f0151e9f4705af7285282f2710e5f756903

File: jans-cedarling/cedarling/src/data/validation.rs (hunk 6 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/validation.rs:232]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/367a7bff7cb93607e338e649fb8eb4114687f476

File: jans-cedarling/cedarling/src/data/mapper.rs (hunk 1 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/4be70739b8df6918ef1bf40c92114c8303cb0900

File: jans-cedarling/cedarling/src/data/mapper.rs (hunk 2 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs:238]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/006de0061734312f3dc03727f06193df2cc64750

File: jans-cedarling/cedarling/src/data/mapper.rs (hunk 3 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/1b5956690d4abf39b98852d56bebbea7d6102180

File: jans-cedarling/cedarling/src/data/validation.rs [hunk 7]
```diff
@@ -0,0 +1,769 @@
...
+
+        // Control characters removed
+        let with_control = json!("hello\x00world");
+        let sanitized = validator.sanitize(&with_control);
+        assert_eq!(sanitized, json!("helloworld"));
+
+        // Null fields removed from objects
+        let with_null = json!({"a": 1, "b": null, "c": 3});
+        let sanitized = validator.sanitize(&with_null);
+        assert_eq!(sanitized, json!({"a": 1, "c": 3}));
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/validation.rs:976]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/a79a332bf273d3d7d1a132c36187b136256afc85

File: jans-cedarling/cedarling/src/data/validation.rs [hunk 8]
```diff
@@ -0,0 +1,769 @@
...
+                        max: self.config.max_object_keys,
+                    });
+                }
+                for (key, val) in obj {
+                    let field_path = format!("{}.{}", path, key);
+                    self.validate_collecting(val, &field_path, depth + 1, errors);
+                }
+            },
+        }
+    }
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/validation.rs:634]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/a79a332bf273d3d7d1a132c36187b136256afc85

File: jans-cedarling/cedarling/src/data/validation.rs [hunk 9]
```diff
@@ -0,0 +1,769 @@
...
+        self.validate_at_path(value, "$", 0)?;
+        Ok(())
+    }
+
+    /// Validate a value and get a detailed result.
+    pub fn validate_detailed(&self, value: &Value) -> ValidationResult {
+        let mut errors = Vec::new();
+        self.validate_collecting(value, "$", 0, &mut errors);
+        ValidationResult::errors(errors)
+    }
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/validation.rs:208]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/a79a332bf273d3d7d1a132c36187b136256afc85

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 10]
```diff
@@ -0,0 +1,711 @@
...
+    fn test_set_nested() {
+        let mapper = CedarValueMapper::new();
+        let mut data = json!({});
+
+        mapper
+            .set_nested(&mut data, "user.profile.name", json!("Alice"))
+            .expect("should set nested value");
+
+        assert_eq!(data, json!({"user": {"profile": {"name": "Alice"}}}));
+    }
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs:930]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/a79a332bf273d3d7d1a132c36187b136256afc85

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 11]
```diff
@@ -0,0 +1,711 @@
...
+                            "type": entity_obj.get("type"),
+                            "id": entity_obj.get("id")
+                        }));
+                    }
+                }
+
+                // Check for __extn marker (extension types)
+                if let Some(extn) = obj.get("__extn") {
+                    return Ok(extn.clone());
+                }
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs:634]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/a79a332bf273d3d7d1a132c36187b136256afc85

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 12]
```diff
@@ -0,0 +1,711 @@
...
+                                "decimal" => Ok(Some(RestrictedExpression::new_decimal(arg))),
+                                "ip" => Ok(Some(RestrictedExpression::new_ip(arg))),
+                                _ => Err(ValueMappingError::InvalidExtensionFormat {
+                                    extension_type: fn_name.to_string(),
+                                    value: arg.to_string(),
+                                }),
+                            };
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs:564]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/a79a332bf273d3d7d1a132c36187b136256afc85

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 13]
```diff
@@ -0,0 +1,711 @@
...
+            Value::Bool(b) => RestrictedExpression::new_bool(*b),
+            Value::Number(n) => {
+                if let Some(i) = n.as_i64() {
+                    RestrictedExpression::new_long(i)
+                } else if let Some(f) = n.as_f64() {
+                    // Convert floating point to decimal extension
+                    RestrictedExpression::new_decimal(f.to_string())
+                } else {
+                    return Ok(None);
+                }
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs:467]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/a79a332bf273d3d7d1a132c36187b136256afc85

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 14]
```diff
@@ -0,0 +1,711 @@
...
+    /// Convert a JSON value to Cedar, returning the inferred Cedar type.
+    ///
+    /// This is useful when you need both the expression and type information.
+    pub fn json_to_cedar_with_type(
+        &self,
+        value: &Value,
+    ) -> Result<Option<(RestrictedExpression, CedarType)>, ValueMappingError> {
+        let cedar_type = CedarType::from_value(value);
+        let expr = self.json_to_cedar(value)?;
+        Ok(expr.map(|e| (e, cedar_type)))
```
[the remark this answers is not in the document]
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs:111]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/a79a332bf273d3d7d1a132c36187b136256afc85

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 15]
```diff
@@ -0,0 +1,1017 @@
...
+    /// - "1h30m45s" (hours, minutes, seconds)
+    /// - "500ms" (milliseconds only)
+    /// - "1d" (days only)
+    fn is_duration_format(value: &str) -> bool {
+        if value.is_empty() {
+            return false;
+        }
+
+        let s = if value.starts_with('-') {
+            &value[1..]
```
Comment (dagregi) [inline jans-cedarling/cedarling/src/data/mapper.rs]: Manually stripping prefix here you can use `strip_prefix()` method

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 16]
```diff
@@ -0,0 +1,1017 @@
...
+
+                RestrictedExpression::new_record(fields)?
+            },
+        };
+
+        Ok(Some(expr))
+    }
+
+    // Estimate the size of a value in bytes
+    fn estimate_value_size(&self, value: &Value) -> usize {
```
Comment (dagregi) [inline jans-cedarling/cedarling/src/data/mapper.rs]: The `self` argument is only used for method calling `estimate_value_size` so it can be removed. Also on the estimations number it would be good to add comments on why those numbers were chosen as estimates

**Comment by olehbozhok** [comment]
Pls, correct me if I'm wrong.  Why do we need the mapper if we need only json value to build a context?
`cedarling/src/authz/build_ctx.rs` 
```
pub(super) fn build_context(
    config: &AuthzConfig,
    request_context: Value,
    build_entities: &BuiltEntities,
    schema: &cedar_policy::Schema,
    action: &cedar_policy::EntityUid,
) -> Result<cedar_policy::Context, BuildContextError> {
```

File: jans-cedarling/cedarling/src/data/mapper.rs [hunk 17]
```diff
@@ -0,0 +1,1017 @@
...
+        // Check for decimal (must contain decimal point and be parseable as f64)
+        // Must have exactly one decimal point and not end with it
+        if value.contains('.') {
+            if value.parse::<f64>().is_ok()
+                && !value.ends_with('.')
+                && value.chars().filter(|&c| c == '.').count() == 1
+            {
+                return Some(ExtensionValue::Decimal(value.to_string()));
+            }
+        }
```
Comment (dagregi) [inline jans-cedarling/cedarling/src/data/mapper.rs:275]: This check will also evaluate values like `1.5e4` which is a scientific notation into a decimal is this maybe a wanted behaviour? sorry if I missed something

**Comment by haileyesus2433** [comment]
> Pls, correct me if I'm wrong. Why do we need the mapper if we need only json value to build a context? `cedarling/src/authz/build_ctx.rs`
> 
> ```
> pub(super) fn build_context(
>     config: &AuthzConfig,
>     request_context: Value,
>     build_entities: &BuiltEntities,
>     schema: &cedar_policy::Schema,
>     action: &cedar_policy::EntityUid,
> ) -> Result<cedar_policy::Context, BuildContextError> {
> ```

i think The mapper should be kept because it provides extension type detection (IP, decimal, datetime, duration) that is essential for populating DataEntry type metadata and validating pushed data is Cedar compatible before storage

File: jans-cedarling/cedarling/src/data/mapper.rs (hunk 17 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs:275]: i think they are considered as a decimal, correct me if am wrong but aren't they a form of decimal representation which makes them a decimal

File: jans-cedarling/cedarling/src/data/mapper.rs (hunk 15 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/49d14e9a24b012eebe979733f1183bdc831fec67

File: jans-cedarling/cedarling/src/data/mapper.rs (hunk 16 above)
Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/data/mapper.rs]: resolved in https://github.com/JanssenProject/jans/pull/13075/commits/6e484dcdd14fabd6a061ec7ab298f70a12dcf2e4

File: jans-cedarling/cedarling/src/data/mapper.rs (hunk 17 above)
Comment (dagregi) [reply jans-cedarling/cedarling/src/data/mapper.rs:275]: Well 1.5e3 is 1500 and 1.5e-3 is 0.0015 so technically it's right so if it parses then it's ok I guess

**Comment by olehbozhok** [comment]
> i think The mapper should be kept because it provides extension type detection (IP, decimal, datetime, duration) that is essential for populating DataEntry type metadata and validating pushed data is Cedar compatible before storage

Ok, anyway, we can delete the unused part in another pr.
Looks ok to me

**Review by dagregi — APPROVED** [review]
